osquery/tables/system/windows/processes.cpp

```diff
@@ -147,238 +214,334 @@ Status getProcList(std::set<long>& pids) {
+  SecureZeroMemory(path.get(), kMaxPathSize);
+  auto ret = QueryFullProcessImageName(proc, 0, path.get(), &out);
+  if (ret != TRUE) {
+    VLOG(1) << "Failed to lookup path information for process " << pid;
+    r["path"] = "-1";
+  } else {
+    r["path"] = TEXT(path.get());
+  }
 
-void genPerfPerProcess(
-    std::map<std::int32_t, std::map<std::string, std::int64_t>>& perfData) {
-  const WmiRequest request(kWinProcPerfQuery);
+  auto s = osquery::pathExists(path.get());
+  r["on_disk"] = INTEGER(s.getMessage());
 
-  if (!request.getStatus().ok()) {
-    VLOG(1) << "Failed to query process perf data from WMI";
-    return;
+  SecureZeroMemory(path.get(), kMaxPathSize);
+  if (pid == GetCurrentProcessId()) {
+    ret = GetModuleFileName(nullptr, path.get(), kMaxPathSize);
+  } else {
+    ret = GetModuleFileNameEx(proc, nullptr, path.get(), kMaxPathSize);
+  }
+  if (ret == FALSE) {
+    VLOG(1) << "Failed to get cwd for " << pid << " with " << GetLastError();
+    r["cwd"] = "-1";
+  } else {
+    r["cwd"] = TEXT(path.get());
   }
+  r["root"] = r["cwd"];
+}
 
-  const auto& results = request.results();
-  for (const auto& result : results) {
-    std::map<std::string, std::int64_t> processData;
-    long processID;
-    long handleCount = 0;
-    std::string elapsedTime;
-    std::string percentProcessorTime;
-
-    result.GetString("ElapsedTime", elapsedTime);
-    result.GetLong("HandleCount", handleCount);
-    result.GetString("PercentProcessorTime", percentProcessorTime);
-    processData["elapsed_time"] = std::stoll(elapsedTime);
-    processData["handle_count"] = handleCount;
-    processData["percent_processor_time"] = std::stoll(percentProcessorTime);
-    result.GetLong("IDProcess", processID);
-    perfData[processID] = processData;
+void genProcessUserTokenInfo(HANDLE& proc, Row& r) {
+  /// Get the process UID and GID from its SID
+  HANDLE tok = nullptr;
+  std::vector<char> tok_user(sizeof(TOKEN_USER), 0x0);
+
+  auto ret = OpenProcessToken(proc, TOKEN_READ, &tok);
+  long elevated = -1;
+  if (ret != 0 && tok != nullptr) {
+    unsigned long tokOwnerBuffLen;
+    ret = GetTokenInformation(tok, TokenUser, nullptr, 0, &tokOwnerBuffLen);
+    if (ret == 0 && GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+      tok_user.resize(tokOwnerBuffLen);
+      ret = GetTokenInformation(
+          tok, TokenUser, tok_user.data(), tokOwnerBuffLen, &tokOwnerBuffLen);
+    }
+
+    /// Check if the process is using an elevated token
+    TOKEN_ELEVATION elevation;
+    DWORD cb_size = sizeof(TOKEN_ELEVATION);
+    if (GetTokenInformation(tok,
+                            TokenElevation,
+                            &elevation,
+                            sizeof(TOKEN_ELEVATION),
+                            &cb_size)) {
+      elevated = elevation.TokenIsElevated;
+    }
+  }
+  r["is_elevated_token"] = INTEGER(elevated);
+  if (ret != 0 && !tok_user.empty()) {
+    auto sid = PTOKEN_OWNER(tok_user.data())->Owner;
```

Comment:
Is this taking data, casting it to a pointer and dereffing? if yes...
![watduck](https://user-images.githubusercontent.com/2386877/48396637-f6eac580-e6cf-11e8-8670-2389f1fadc97.jpg)

Reply:
Yes, yes it is :) I’ll make it a bit safer.